Defer delivering port_status received during handshake

samueljero opened the pull request:
POX currently drops port_status messages during the switch--controller handshake.
Unfortunately this leads to a race condition where important topology changes
can occur between the time that the features_reply message is received--
containing a complete list of ports at the moment it was sent--and the time
the handshake completes and port_status messages begin to be delivered.

This patch collects port_status messages received on a connection prior to the
switch--controller handshake completing and defers processing them, and sending
port_status events, until the connection completes.

Closes issue #162

samueljero committed:
Defer delivering port_status received during handshake

POX currently drops port_status messages during the switch--controller handshake.
Unfortunately this leads to a race condition where important topology changes
can occur between the time that the features_reply message is received--
containing a complete list of ports at the moment it was sent--and the time
the handshake completes and port_status messages begin to be delivered.

This patch collects port_status messages received on a connection prior to the
switch--controller handshake completing and defers processing them, and sending
port_status events, until the connection completes.

MurphyMc commented:
Ugh.  The fact that anything is fair game before handshaking is complete continues to make me sad!

Good find.  We should definitely fix this, and without having looked at it _too_ hard, your approach seems good.  See nitpicky reviews. :)

samueljero committed:
Updates as a result of review

1)Use con.info() instead of log.debug()
2)Change variable name to fix spelling and indicate POX-internal
3)Set con._deferred_port_status to None at end

samueljero commented:
Well, OpenFlow doesn't really specifiy any of this handshaking (other than of_hello). We just need this handshake of messages to do much of anything useful :) But yes, would have been nice if they had specified a more useful handshake.

I've updated the pull request taking into account your comments.

MurphyMc commented:
So I've got a modified version of this pretty much ready to go.  The difference is that it only buffers port status messages between the features reply and the finishing of the handshake, since any that come _before_ the features reply should hopefully be superseded by the port info in the features reply.

Essentially, the deferred list starts as None, and is set to an empty list when the features reply comes.  When receiving a port status, it's appended to the list only if it's not None.

I'm not sure if you're still working on this, but it'd be great if you could confirm that this resolves your issue (that is, that it's really the loss of port status messages between the features reply and the connection going up that are problematic, and not ones before the features reply).

MurphyMc commented:
Hopefully addressed by 0cc9de5d2bc95ebc1cf232b21d788ba931faa6d0
See additional discussion in #162
